std/stdio.d

```diff
@@ -2907,25 +2918,39 @@ is empty, throws an `Exception`. In case of an I/O error throws
             static if (c.sizeof == 1)
             {
                 // simple char
+                rbuf16ShouldBeEmpty();
                 if (orientation_ <= 0) trustedFPUTC(c, handle_);
                 else trustedFPUTWC(c, handle_);
             }
             else static if (c.sizeof == 2)
             {
                 import std.utf : encode;
+                import std.utf : decodeFront;
 
                 if (orientation_ <= 0)
                 {
                     if (c <= 0x7F)
                     {
+                        rbuf16ShouldBeEmpty();
                         trustedFPUTC(c, handle_);
                     }
-                    else
+                    else if (0xD800 <= c && c <= 0xDBFF) // high surrogate
                     {
-                        char[4] buf;
-                        immutable size = encode(buf, c);
-                        foreach (i ; 0 .. size)
-                            trustedFPUTC(buf[i], handle_);
+                        rbuf16ShouldBeEmpty();
+                        rbuf16[0] = c;
+                        rbuf16Filled = 1;
+                    }
+                    else // standalone or low surrogate
+                    {
+                        rbuf16[rbuf16Filled] = c;
+                        ++rbuf16Filled;
+                        wchar[] str = rbuf16[0 .. rbuf16Filled];
+                        rbuf16Filled = 0;
+                        immutable dchar d = decodeFront(str);
+                        char[4] wbuf;
+                        immutable size = encode(wbuf, d);
+                        foreach (i; 0 .. size)
+                            trustedFPUTC(wbuf[i], handle_);
```

Comment:
This seems like a lot of acrobatics for this purpose. A few comments here:

1. You only need a buffer of 1 wchar.
2. You don't need a "size" for checking if there are 0 or 1 wchars in there, just store a `wchar(0)` when you aren't using it.
3. From my experience with iopipe, you get better performance when you avoid using members to do the decoding.  This would be better off being duplicated in the function which takes whole wstrings, and you can then use a local to do all the decoding/encoding.

Reply:
Addressed points 1 and 2. Hopefully, the code's clearer now.

I think I'll leave point 3 for another PR.

Reply:
Good progress on the simplification. One more thing, I thought it might look better actually if you handle the surrogate pair explicitly instead of the ?: operator
```D
else
{
   dchar d = c; // default to just translate directly
   if(highSurrogate)
   {
      immutable wchar[2] buf = [highSurrogate, c];
      d = buf[].front; // doesn't this just work? Do we need to use decodeFront?
   }
   // .. all the rest of the stuff you have
}
```

Reply:
It would be cool if std.utf provided a way to decode N code units directly. For example, if `decode(highSurrogate, c)` just worked.

Reply:
Your code misses the case where c is an unpaired low surrogate. But encode throws on that, so that works. Done.